### PR TITLE
Add ValuePattern.Value changed event to TextBox

### LINKED ISSUE #10257: ValuePattern.Value property changed event is not sent from TextBox control.
### .NET version

.NET 8.0 RC2

### Did it work in .NET Framework?

Yes

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

.NET 7.0.13

### Issue description

When listening to UIA events from the TextBox control, only text changed event is sent in .NET 8. However, there is also value property changed event in .NET 7.

.NET 7
https://github.com/dotnet/winforms/assets/85718983/6fbb2ba7-083a-4185-8662-a411fe9d2fcd

.NET 8
https://github.com/dotnet/winforms/assets/85718983/fd75b565-f281-4a48-aa00-1fe1ac568c22

This may be caused by this PR https://github.com/dotnet/winforms/pull/7741 that implements the UIA provider for TextBox.

### Steps to reproduce

1. Insert a TextBox control to a Form.
2. Run the application.
3. Open Accessibility Insights for Windows and listen to UIA events of the TextBox.
4. Input characters in the TextBox and observe the events.

### PR DESCRIPTION
**Pull request by SimonZhao888**
Fixes #10257 

## Proposed changes

- Add raise ValuePattern.Value property changed event to TextBoxBase.
- Override Name property of TextBoxBaseAccessibleObject

## Customer Impact

- ValuePattern.Value property changed event will be sent from TextBox control.
- When we don't set the accessible name of the textbox or the previous label, the accessible name of the textbox will return "" not the value of textbox.

## Regression? 

- Yes 

## Risk

- Minimal

### Before
ValuePattern.Value property changed event is not sent from TextBox.
image

### After
When we make changes to the text in the textbox, it will send the ValuePattern.Value property changed event.
TextBox_ValuePattern1

## Test methodology 

- Manual (Insights)

## Test environment(s) 

- 9.0.100-alpha.1.23504.14

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/10295)

### CODE REVIEW

**Review by ricardobossan — COMMENTED** [review]
LGTM

**Review by lonitra — DISMISSED** [review]
LGTM, does this need to be serviced?

**Comment by SimonZhao888** [comment]
Hi @Tanya-Solyanik, do we need merge this change to Release/8.0?

**Review by Tanya-Solyanik — DISMISSED** [review]
@SimonZhao888  - I assume this regression was introduced by https://github.com/dotnet/winforms/pull/7741

**Review by Tanya-Solyanik — APPROVED** [review]
Thank you!

**Comment by SimonZhao888** [comment]
Hi @Tanya-Solyanik ,

I have a question that I don't understand, when we don't set the accessible name of the textbox or the previous label, the accessible name of the textbox will return Null (https://github.com/dotnet/winforms/blob/24c362b81da76ada1d80cbbabafa0da3b0bfc6c2/src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlAccessibleObject.cs#L291-L301). However, in the Accessible tools, the Name is assigned the value of textBox(this happens only when Name = null). 

We want to know at what point this action of assigning a value takes place. We didn't find the codes of update the value of Name after the Name return null.

Thanks,
Simon

**Comment by Tanya-Solyanik** [comment]
>We want to know at what point this action of assigning a value takes place. We didn't find the codes of update the value of Name after the Name return null.

I assume this is done by the MSAA proxy in UIAccessibilityCore. If name is not set by the application, native accessibility code attempts to find the name based on other data.

**Comment by SimonZhao888** [comment]
These failed checks were not caused by this PR.

**Comment by Tanya-Solyanik** [comment]
>These failed checks were not caused by this PR.

Hm, but tests are passing for other PRs

File: src/System.Windows.Forms/src/System/Windows/Forms/WinFormsUtils.cs [hunk 1]
```diff
@@ -244,7 +243,7 @@ public static char GetMnemonic(string? text, bool convertToUpperCase)
     {
         if (text is null)
         {
-            return null;
+            return string.Empty;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/WinFormsUtils.cs:246]: This change is a breaking one. `null` means that the UIACore will continue looking for the name while `Empty` string is used when an empty name is desired.

Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/WinFormsUtils.cs]: Yes, I think I made a cheap mistake. I've withdrawn this change, and update Name property of TextBoxBaseAccessibleObject.cs class to fix the regression issue.

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBox/TextBoxBase.cs [hunk 2]
```diff
@@ -1583,6 +1584,8 @@ private protected virtual void RaiseAccessibilityTextChangedEvent()
         if (IsAccessibilityObjectCreated)
         {
             AccessibilityObject.RaiseAutomationEvent(UIA_EVENT_ID.UIA_Text_TextChangedEventId);
+            using var textVariant = PasswordProtect ? VARIANT.Empty : (VARIANT)Text;
```
Comment (lonitra) [inline src/System.Windows.Forms/src/System/Windows/Forms/TextBox/TextBoxBase.cs:1587]: Curious why we changed from `(VARIANT)string.Empty` -> `VARIANT.Empty`?

**Review by lonitra — COMMENTED** [review]
Looks good to me, just a question. Please also resolve conflicts.

File: src/System.Windows.Forms/src/System/Windows/Forms/WinFormsUtils.cs (hunk 1 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/WinFormsUtils.cs]: Is it possible to avoid making this change?

Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/WinFormsUtils.cs]: I've reversed the change.

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBox/TextBoxBase.cs (hunk 2 above)
Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBox/TextBoxBase.cs:1587]: Varint.Empty, if used in the context of a string, indicates a zero-length string.

Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBox/TextBoxBase.cs:1587]: Sure, but `(VARIANT)string.Empty` is a more self-documenting code.

Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBox/TextBoxBase.cs]: Good advice, I've updated it.

**Review by lonitra — APPROVED** [review]

**Comment by Tanya-Solyanik** [comment]
@SimonZhao888  - had this change been tested by Olina's team?

**Comment by SimonZhao888** [comment]
> @SimonZhao888 - had this change been tested by Olina's team?

Yes, this has been tested by Olina's team.

**Comment by SimonZhao888** [comment]
> @SimonZhao888 - had this change been tested by Olina's team?

Also, I'll ask the Olena team to test PR [10369](https://github.com/dotnet/winforms/pull/10369) as well.

**Pull request merged**